src/main/kotlin/fr/xebia/xke/5_FP.kt

```diff
@@ -0,0 +1,71 @@
+package fr.xebia.xke
+
+import arrow.core.*
+import arrow.syntax.option.toEither
+
+import arrow.core.Either
+import arrow.core.Option
+import arrow.core.flatMap
+import arrow.data.NonEmptyList
+import arrow.data.Validated
+import arrow.data.applicative
+import arrow.data.ev
+
+/**
+ * This class validates user data in two ways:
+ * - Applicative for all fields (validated all fields)
+ * - Monadic within each field (stops on the first error per field)
+ */
+class UserValidation {
+
+    /*
+     * The following functions might help you
+     * - Option.fromNullable("something")
+     * - Option(maybeRight).toEither { "Left value" }
+     * - Either.cond(true, { 42 }, { "Error" })
+     * - Validated.fromEither(...)
+     * - Validated.applicative<List<String>>()
+            .map2(validated1, validated2, { v: Tuple<Int, Int> -> v.a + v.b })
+            .ev()
+     * - NonEmptyList.of("a string", "another string")
+     */
+
+    // TODO find a way to aggregate in an applicative way the 'readName' and 'readAge' functions
+    fun validateUser(params: Map<String, String>): Validated<NonEmptyList<String>, User>  {
+        return Validated.applicative<NonEmptyList<String>>()
+            .map2(
```

Comment:
- Don't understand how we can extend this mechanism to 3 or more checks.
- Here we have a List of errors, ok, but each validation `readName` and `readAge` should only return one error not a list of errors. Typing should provide an hint about expected behaviour.

Reply:
I'm not inventing nothing new here!
this is the way it is typically done in several FP languages

Reply:
No I haven't, I was asking how we could apply check to 3 or more fields :)
Use of `applicative.tupled` in the last comment answers this question